docs/adr/0002-database-corruption.md

```diff
@@ -0,0 +1,132 @@
+# Handling Database Corruption
+
+* Status: accepted
+* Date: 2021-06-08
+
+## Context and Problem Statement
+
+Some of our users have corrupt SQLite databases and this makes the related
+component unusable.  The best way to deal with corrupt databases is to simply
+delete the database and start fresh (see #2628).  However, it's not so easy
+to know if a database operation error is from a corrupt database or from bad
+code.  We want a system that identifies and deals with database corruption but
+avoids deleting data based on logic errors.  That system needs to deal with 3
+related questions:
+
+  * A) When do we identify corrupted databases?
+  * B) How do we identify corrupted databases?
+  * C) What do we do when we identify corrupted databases?
+
+## Decision Drivers
+
+* Deleting valid user data should be avoided at almost any cost
+* Keeping a corrupted database around is almost as bad.  It currently prevents
+  the component from working at all.
+* We don't currently have a good way to distinguish between the two error
+  cases, but this can be improved by reviewing telemetry and sentry data.
+
+## Considered Options
+
+* A) When do we identify corrupted databases?
+  * At migration time only
+  * Whenever we perform a database operation
+* B) How do we identify corrupted databases?
+  * Assume all errors during migration are from corrupt databases
+  * Compare error values against known corruption errors for all operations
+  * Compare error values against known corruption errors when opening the database
+* C) What do we do when we identify corrupted databases?
+  * Delete the database file and recreate the database
+  * Move the database file and recreate the database
+  * Have the component fail
+
+## Decision Outcome
+
+* A: Whenever we perform a database operation
+
+This was chosen because it can be a long time between migrations and we want
+to de able to fix corrupt database during that period.
+
+* B: Compare error values against known corruption errors when opening the database
+
+This was chosen because it had the simplest initial implementation and was
+least likely to misidentify errors.  We can target only errors that clearly
+indicate corruption, for example the `DatabaseCorrupt` error code.  We can also
+only target opening the database, which is the easiest time to fix issues.
+
+* C: Delete the database file and recreate the database
+
+This follows the choice of B.  Since we're being conservative in identifying
+errors, we can delete the database file with relative confidence.
+
+### Implementation notes:
+
+This can be implemented by updating the `Store` code to wrap each database
```

Comment:
and here we are back to "each database operation". I think this is going to be difficult in practice as (a) the code detecting the error doesn't have access to all connections to close and (b) there's a risk apps will behave badly in that scenario - eg, the strategy here might be to (say) somehow flag the detection and delete at next open?

However, given this should be long-lived, I don't think we are actually making the decision that we will never check on each operation - it's closer to "we want to be cautious, so will *initially* start with just on open, but will improve this as we learn.

Or to put it another way, if we do end up implementing the "every operation" strategy later, this adr should remain accurate without changes.

Reply:
Oops, I meant to delete this one with my last edit.

I agree that there's definitely some risk, but the more I think about it the more it seems manageable.  For (a) we would make it so only the `Store` has access to the raw connection and the other components access the DB through the store.  Then the store should be free to drop the connection and create a new one.  For (b) it feels like apps will behave badly if you drop all the data on them, but I'm struggling to come up with something concrete.  For example: I fetch a set of logins, do something with them, then run another query later on.  Even today, I can't assume that any of the logins are still in the database -- they may have been deleted by a sync.  Also, the current process is to return an error, which seems about as likely to me to cause issues.

In any case, I definitely agree with the decision to initially target with opening the DB, learn more and then move to other operations.